pkg/rbac/sharedData.go

```diff
@@ -32,116 +40,173 @@ type resource struct {
-				if err != nil {
-					klog.Errorf("Error %s retrieving rows for query:%s for apigroup %s and kind %s", err.Error(), query,
-						apigroup, kind)
-					continue
-				}
+// Obtain all the namespaces in the hub cluster.
+// Equivalent to `oc get namespaces`
+func (shared *SharedData) GetSharedNamespaces(cache *Cache, ctx context.Context) error {
+	shared.nsLock.Lock()
+	defer shared.nsLock.Unlock()
+	//empty previous cache
+	shared.namespaces = nil
+	shared.nsErr = nil
 
-				shared.csResources = append(shared.csResources, resource{apigroup: apigroup, kind: kind})
+	klog.V(5).Info("Getting namespaces from Kube Client..")
 
-			}
-		}
+	clientset, kubeErr := kubernetes.NewForConfig(cache.restConfig)
+	if kubeErr != nil {
+		klog.Warning("Error with creating a new clientset.", kubeErr.Error())
+		shared.nsErr = kubeErr
+		shared.nsUpdatedAt = time.Now()
+		return shared.nsErr
+	}
 
-		// get all namespaces in cluster and cache in shared.namespaces.
-		shared.nsLock.Lock()
-		defer shared.nsLock.Unlock()
-		allNamespaces, nsErr := shared.GetSharedNamespaces(cache, ctx)
-		if nsErr != nil {
-			shared.nsErr = nsErr
-		}
+	namespaceList, nsErr := clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if nsErr != nil {
+		klog.Warning("Error resolving namespaces from KubeClient: ", nsErr)
+		shared.nsErr = nsErr
+		shared.nsUpdatedAt = time.Now()
+		return shared.nsErr
+	}
 
-		// update the cache.
-		shared.namespaces = allNamespaces
-		shared.csUpdatedAt = time.Now()
-	} else {
-		klog.V(8).Info("Using cluster scoped resources from cache.")
+	// add namespaces to allNamespace List
+	for _, n := range namespaceList.Items {
+		shared.namespaces = append(shared.namespaces, n.Name)
 	}
+	shared.nsUpdatedAt = time.Now()
 
-	return shared.csResources, shared.csErr
+	return shared.nsErr
 }
 
-func (shared *SharedData) GetSharedNamespaces(cache *Cache, ctx context.Context) ([]string, error) {
+func (shared *SharedData) GetSharedManagedClusters(cache *Cache, ctx context.Context) error {
 
-	if len(shared.namespaces) > 0 &&
-		time.Now().Before(shared.nsUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Millisecond)) {
-		klog.V(5).Info("Using namespaces from shared cache")
+	shared.mcLock.Lock()
+	defer shared.mcLock.Unlock()
+	// clear previous cache
+	shared.managedClusters = nil
+	shared.mcErr = nil
 
-	} else {
+	var managedClusters []string
 
-		klog.V(5).Info("Getting namespaces from Kube Client..")
+	scheme := runtime.NewScheme()
+	scheme.AddKnownTypes(managedClusterResourceGvr.GroupVersion())
 
-		cache.restConfig = config.GetClientConfig()
-		clientset, err := kubernetes.NewForConfig(cache.restConfig)
-		if err != nil {
-			klog.Warning("Error with creating a new clientset.", err.Error())
+	resourceObj, err := cache.dynamicClient.Resource(managedClusterResourceGvr).List(ctx, metav1.ListOptions{})
 
-		}
+	if err != nil {
+		klog.Warning("Error resolving resources with dynamic client", err.Error())
+		return shared.mcErr
+	}
```

Comment:
Save the error and time in the cache.
```suggestion
	if err != nil {
		klog.Warning("Error resolving resources with dynamic client", err.Error())
		shared.mcErr = err
		shared.mcUpdatedAt = time.Now()
		return shared.mcErr
	}
```